shared/src/test/scala/AwaitBehavior.scala

```diff
@@ -100,21 +99,26 @@ class AwaitBehavior extends munit.FunSuite:
       f.await
 
     val f = Future:
-      var loop = true
-      while loop && !Async.current.group.isCancelled do
-        // on scnative, suspending and changing the carrier thread currently kills lock monitors
-        reqCh.readSource.poll().map(_.right.get).foreach {
-          case ReqMessage.Lock =>
-            resCh.sendImmediately(ResMessage.LockResult(l.acquireLock()))
-          case ReqMessage.Release =>
-            l.releaseLock()
-            resCh.sendImmediately(ResMessage.Done)
-          case ReqMessage.Complete(data) =>
-            l.complete(data, s)
-            resCh.sendImmediately(ResMessage.Done)
-          case ReqMessage.Quit => loop = false
-          case _               => ??? // does not happen
-        }
+      // on scnative, suspending and changing the carrier thread currently kills lock monitors
+      // so we run the body in a special single-threaded context
+      Async.blocking {
+        var loop = 0
+        while loop >= 0 && !Async.current.group.isCancelled do
+          if loop == 0 then AsyncOperations.`yield`()
+          loop = (loop + 1) % 10
+          reqCh.readSource.poll().map(_.right.get).foreach {
+            case ReqMessage.Lock =>
+              resCh.sendImmediately(ResMessage.LockResult(l.acquireLock()))
+            case ReqMessage.Release =>
+              l.releaseLock()
+              resCh.sendImmediately(ResMessage.Done)
+            case ReqMessage.Complete(data) =>
+              l.complete(data, s)
+              resCh.sendImmediately(ResMessage.Done)
+            case ReqMessage.Quit => loop = -1
+          }
+      }(using SingleThreadedSupport)
```

Comment:
Perhaps moving the `using` clause of `Async.blocking` to _before_ the `body` will make this look much nicer